lib/RestructureCFG/CycleEquivalenceAnalysis.cpp

```diff
@@ -0,0 +1,659 @@
+      if (Label->type() == EdgeType::Invalid) {
+        Label->setType(EdgeType::BackEdge);
+      }
+    }
+  }
+}
+
+/// Helper function to materialize the capping backedges
+template<class GraphT, class GT>
+void CycleEquivalenceAnalysis<GraphT, GT>::insertCappingBackedges(BlockGraph
+                                                                    &Graph) {
+  // Highest, i.e., closer to the spanning tree root, reached DFSNum for each
+  // `BB`
+  std::map<BlockNode *, size_t> HighestReachedDFSNum;
+
+  // Keep a mapping between the `DFSNum` and the corresponding node
+  std::map<size_t, BlockNode *> DFSNumNodeMap;
+
+  llvm::SmallVector<BlockNode *> PostOrder;
+  for (BlockNode *BB :
+       llvm::post_order(llvm::Undirected<BlockNode *>(Graph.getEntryNode()))) {
+    PostOrder.push_back(BB);
+  }
+
+  for (BlockNode *BB : PostOrder) {
+
+    // 1) Compute the highest reached node from a backedge starting from the
+    //    current `Block`
+    size_t HI0 = SizeTMaxValue;
+
+    // We are only interested in looking at backedges
+    for (const auto &[BackwardEdge, _, __] : getBackwardEdges(BB)) {
+      BlockNode *Target = std::get<1>(BackwardEdge);
+      HI0 = std::min(HI0, Target->getDFSNum());
+
+      // We also populate the mapping between the `DFSNum` and the node
+      DFSNumNodeMap[Target->getDFSNum()] = Target;
+    }
+
+    // 2) Compute the highest reached node for any child of the current `Block`
+    size_t HI1 = SizeTMaxValue;
+    for (const auto &[ForwardEdge, _, __] : getForwardEdges(BB)) {
+      BlockNode *Target = std::get<1>(ForwardEdge);
+      HI1 = std::min(HI1, HighestReachedDFSNum.at(Target));
+    }
+
+    // 3) Assign the highest reached node for the current `Block`
+    HighestReachedDFSNum[BB] = std::min(HI0, HI1);
+
+    // 4) Compute the child which has the reached DFSNum computed at step 2)
+    BlockNode *HI1Child = nullptr;
+    if (HI1 != SizeTMaxValue) {
+      for (const auto &[ForwardEdge, _, __] : getForwardEdges(BB)) {
+        BlockNode *Target = std::get<1>(ForwardEdge);
+        if (HI1 == HighestReachedDFSNum.at(Target)) {
+          HI1Child = Target;
+          break;
+        }
+      }
+    }
+
+    // 5) This is the highest reached node for any child of the current `Block`,
+    //    excluding the one which reaches the highest target
+    size_t HI2 = SizeTMaxValue;
+    for (const auto &[ForwardEdge, _, __] : getForwardEdges(BB)) {
+      BlockNode *Target = std::get<1>(ForwardEdge);
+      if (Target != HI1Child) {
+        HI2 = std::min(HI2, HighestReachedDFSNum.at(Target));
+      }
+    }
+
+    // Insert the capping backedge, if the second highest reached node for any
+    // child of the current `BB`, is higher than any backedges which starts from
+    // `BB`. The capping backedge goes from `BB` to such second highest reached
+    // node.
+    if (HI2 < HI0) {
+      BlockNode *HI2Node = DFSNumNodeMap.at(HI2);
+
```

Comment:
The only reason why you need `DFSNumNodeMap` is to know the node with `DFSNum == HI2`. But you've just computed `HI2`, and you knew exactly what that node was.

So, drop `DFSNumNodeMap` an rewrite the loop at point `5)` as:
```c++
    size_t HI2 = SizeTMaxValue;
    BlockNode *HI2Node = nullptr;
    for (const auto &[ForwardEdge, _, __] : getForwardEdges(BB)) {
      BlockNode *Target = std::get<1>(ForwardEdge);
      if (Target != HI1Child) {
        size_t NewMin = std::min(HI2, HighestReachedDFSNum.at(Target));
        if (NewMin < HI2)
          HI2Node = Target;
        HI2 = NewMin;
      }
    }
```

Reply:
Mmmmm, I think you got the functioning of this wrong. When I compute the HI2 `DFSNum` (the number of the second highest reached node for each child of the current `BB` node), HI2 does not refer to the `Target` nodes on which I iterate on.

The procedure works like this:
- I iterate on the forward edges from `BB`. The other "end" of such edges will be the `Target` node. Then, using the `HighestReachedDFSNum` mapping, I retrieve the "highiest" `DFSNum` that is reached from each subtree starting from `Target` which has been previously computed. Such highest `DFSNum` does not correspond to `Target`, but can be a node which is arbitrarily "high" in the tree (in other words, there is a "preliminary" phase which for each node, computes the highest reached node by a backedge, which populates the `HighestReachedDFSNum` mapping. Then, in this phase, for each child `Target` of the current `BB`, I select the highest reached `DFSNum` using the mapping).
- The `DFSNum` is then passed in the `DFSNumNodeMap` in order to retrieve the node to which we need to add the capping backedge to.

We can get rid of  `DFSNumNodeMap`, at the cost of saving directly the pointer to `HI2Node` together with the `HI2` `DFSNum` in the `HighestReachedDFSNum`. Let me know if you would prefer things to be done in such way.

Reply:
Yes please

Reply:
I implemented the proposed changes from [here](https://github.com/revng/revng-c/blob/f23593da9f15288aca2fee41975e6e55dbbbcb3b/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L211) (the main change is that now the `HighestReachedDFSNum` contains both the `DFSNum` (which is needed to apply the `<` comparisons), and the pointer to the original node.

I think the code is now even less understandable (since it deviates from the paper description), so I also added some comments before each of the 4 steps of the capping backedges introduction, to better explain what the variables represents, hopefully making it easier to reconstruct the functioning of this procedure.

Reply:
I checked it out. I kinda like it, but I wouldn't use a `std::pair` in the map, I'd use a custom `struct` locally defined in the function with fields with explanatory naming.

Besides that I'd say the current implementation is fine even if it deviates from the paper

Reply:
Oh ok, I'll use a custom `struct` in place of the `pair`

Reply:
Ok, [here](https://github.com/revng/revng-c/blob/09b768691c58208a87bb2184f9741d98110dc992/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L209) the custom `struct` which represents the `BlockNode` and its `DFSNum`